### PR TITLE
Improve __spreadArray perf, and other fixes related to SpreadElement

### LINKED ISSUE #44287: Performance regression in `__spreadArray`
# Bug Report

### 🔎 Search Terms

__spreadArray performance benchmarking

### 🕗 Version & Regression Information

- This changed between versions 4.1 and 4.2

### ⏯ Playground Link

[CodeSandbox with benchmarks](https://codesandbox.io/s/old-https-xrgng)
[Playground link with configuration example](https://www.typescriptlang.org/play?target=1&ts=4.2.3#code/MYewdgzgLgBAZiEMC8MDaBGANDATDgZgF0BuAKFElgCMBDAJxXQBYcBWHANlIvGnkR1GqNADpxCEDnGihRIA)

### 💻 Code

```ts
__spreadArray(__spreadArray([], arr1), arr2)
// vs.
__spreadArrays(arr1, arr2)
```

### 🙁 Actual behavior

Starting from 4.2, occurences of `__spreadArrays` are replaced with multiple calls to `__spreadArray`. The change fixes side effects in the polyfill (#32959).

| Function     | Browser    | Arrays     | Elements per Array | Runs | Mean time in ms | Slowdown |
| :------------- | :----------: | -----------: | -----------: | -----------: |-----------: | -----------: |
| __spreadArrays | Chromium 90 | 1 | (numbers) 1000000 | 100 | 2.32 | |
| __spreadArray | Chromium 90 | 1 | (numbers) 1000000 | 100 | 10.29 | 4.43 |
| __spreadArrays | Chromium 90 | 2 | (numbers) 1000000 | 100 | 4.35 | |
| __spreadArray | Chromium 90 | 2 | (numbers) 1000000 | 100 | 25.4 | 5.83 |
| __spreadArrays | Chromium 90 | 2 | (objects) 1000000 | 100 | 4.82 | |
| __spreadArray | Chromium 90 | 2 | (objects) 1000000 | 100 | 68.1 | 14.10 |

### 🙂 Expected behavior

No or moderate performance impact.

### PR DESCRIPTION
**Pull request by rbuckton**
This changes `__spreadArray` to use `Array.prototype.concat` and avoids explicitly converting a sparse array to a packed array when such a conversion is unnecessary or unexpected. This also fixes an issue where `[1, , ...a]` would drop the missing element at index 1.

Before this can be merged I need to verify backwards compatibility with `__spreadArray` for older versions of TS. I suspect I may need to change the `if` condition to be backwards compatible (where we always packed the result).

Fixes #44287

### CODE REVIEW

**Comment by rbuckton** [comment]
@typescript-bot perf test

**Comment by rbuckton** [comment]
I don't expect to see much of a perf change since our own compiler doesn't use spread very often. There may be a bump due to adding a new literal value argument to each spread. If so I can investigate adding a reusable and immutable `true`/`false` literal.

@typescript-bot perf test
@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by rbuckton** [comment]
@typescript-bot perf test

**Comment by rbuckton** [comment]
Running one more set of perf tests. The previous test was against `master`, which is over a week old since the switch to `main`. Now that I've updated the benchmark framework to point to `main`, I should have more accurate results.

@typescript-bot perf test

**Comment by rbuckton** [comment]
That's much better. I was concerned that the material-ui tests looked slower, but it was due to diffing against an old baseline.

We need to merge and ship https://github.com/microsoft/tslib/pull/151 before I can merge this, unless I remove the checker restriction for `__spreadArray` here: https://github.com/microsoft/TypeScript/pull/44527/files#diff-d9ab6589e714c71e657f601cf30ff51dfc607fc98419bf72e04f6b0fa92cc4b8R40729

I'd prefer the restriction remain so that we have a way to indicate to the user they need to update their copy of `tslib` to support this change.

**Pull request merged**
